docs/spec/lightclient/detection/req-ibc-detection.md

```diff
@@ -0,0 +1,296 @@
+  trusting period) a correct relayer checks whether the handler is on a
+  different branch than the relayer. 
+  And we would like that this is enough to achieve
+  the Handler requirement.
+  
+   - here the correctness argument would be easy if a correct relayer is
+     based on a light client with a *trusted* state, that is, a light
+     client who never changes its opinion about trusted. Then if such a
+     correct relayer checks-in with a handler, it will detect a fork, and
+     act in time.
+
+   - if the light client does not provide this interface, in the case of
+     a fork, we need some assumption about a correct relayer being on a
+     different branch than the handler, and we need such a relayer to 
+	 check-in not too late. Also
+     what happens if the relayer's light client is forced to roll-back
+     its lightstore?
+     Does it have to re-check all handlers?
+
+
+
+
+## On the interconnectedness of things
+
+In the broader discussion of so-called "fork accountability" there are
+several subproblems
+
+- Fork detection 
+
+- Evidence creation and submission
+
+- Isolating misbehaving nodes (and report them for punishment over abci)
+
+
+### Fork detection
+
+The preliminary specification ./detection.md formalizes the notion of
+a fork. Roughly, a fork exists if there are two conflicting headers
+for the same height, where both are supported by bonded full nodes
+(that have been validators in the near past, that is, within the
+trusting period). We distinguish between *fork on the chain* where two
+conflicting blocks are signed by +2/3 of the validators of that
+height, and a *light client fork* where one of the conflicting headers
+is not signed by  +2/3 of the current height, but by +1/3 of the
+validators of some smaller height.
+
+In principle everyone can detect a fork
+
+- ./detection talks about the Tendermint light client with a focus on 
+  light nodes.
+  
+- a relayer runs such light clients and may detect forks in this way
+
+- in IBC, a relayer can see that a handler is on a conflicting branch
+    - the relayer should feed the handler the necessary information so
+      that it can halt
+	- the relayer should report the fork to a full node
+	
+### Evidence creation and submission
+
+- the information sent from the relayer to the handler could be called
+  evidence, but this is perhaps a bad idea because the information sent to a
+  full node can also be called evidence. But this evidence might still
+  not be enough as the full node might need to run the "fork
+  accountability" protocol to generate evidence in the form of
+  consensus messages. So perhaps we should
+  introduce different terms for:
+  
+  - proof of fork for the handler (basically consisting of lightblocks)
+  - proof of fork for a full node (basically consisting of (fewer) lightblocks)
+  - proof of misbehavior (consensus messages)
+  
+
+
+### Isolating misbehaving nodes
+
+- this is the job of a full node. 
```

Comment:
I think that we have some missing parts here: a full node should verify proof of fork and if valid, it should create the corresponding evidence that should be published on chain using evidence submission protocol. There are several kind of evidences and they don't have the same treatment on chain. Double signing and lunatic are self-contained and they can be processed without additional information. Evidence for amnesia requires running challenge-response protocol (posting votesets on chain by suspected validators) and running fork accountability procedure to catch bad validators. We need to precisely specify proof of fork verification logic and also how valid proof is turned into evidence(s).

Reply:
Yes. You are right. I drew decision tree while discussing with @ebuchman but didn't add it in the draft. I overlooked that. I will add that!

Eventually the question is where this should go, because at the end, this is independent of all the light client and IBC related questions, but just needs proof of fork as input.

Reply:
That's right. It's actually Tendermint specific stuff.